core/domain/prod_validation_jobs_one_off.py

```diff
@@ -2474,213 +2600,1326 @@ def _get_custom_validation_functions(cls):
+                continue
+            story_rights = story_services.get_story_rights(story_model.id)
+            if not story_rights.story_is_published:
+                cls.errors['public story check'].append(
+                    'Entity id %s: Story with id %s corresponding to entity '
+                    'is private' % (item.id, story_model.id))
+
+    @classmethod
+    def _validate_completed_nodes(cls, item):
+        """Validates that completed nodes belong to the story
+
+        Args:
+            item: ndb.Model. StoryProgressModel to validate.
+        """
+        completed_activity_model = user_models.CompletedActivitiesModel.get(
+            item.user_id)
+        story_model_class_model_id_model_tuples = (
+            cls.external_instance_details['story_ids'])
+        for (_, _, story_model) in (
+                story_model_class_model_id_model_tuples):
+            # The case for missing story external model is ignored here
+            # since errors for missing story external model are already
+            # checked and stored in _validate_external_id_relationships
+            # function.
+            if story_model is None or story_model.deleted:
+                continue
+            story_node_ids = [
+                node['id'] for node in story_model.story_contents['nodes']]
+            invalid_node_ids = [
+                node_id
+                for node_id in item.completed_node_ids if node_id not in (
+                    story_node_ids)]
+            if invalid_node_ids:
+                cls.errors['completed node check'].append(
+                    'Entity id %s: Following completed node ids %s do '
+                    'not belong to the story with id %s corresponding '
+                    'to the entity' % (
+                        item.id, invalid_node_ids, story_model.id))
+
+            # Checks that the explorations corresponding to completed nodes
+            # exist, are marked as completed in CompletedActivitiesModel
+            # and are public.
+            private_exp_ids = []
+            missing_exp_ids = []
+            unmarked_exp_ids = []
+            for node in story_model.story_contents['nodes']:
+                if node['id'] in item.completed_node_ids:
+                    exp_id = node['exploration_id']
+                    exp = exp_models.ExplorationModel.get_by_id(exp_id)
+                    if exp is None or exp.deleted:
+                        missing_exp_ids.append(exp_id)
+                        continue
+                    if exp_id not in completed_activity_model.exploration_ids:
+                        unmarked_exp_ids.append(exp_id)
+                    if rights_manager.is_exploration_private(exp_id):
+                        private_exp_ids.append(exp_id)
+            error_msg = ''
+            if private_exp_ids:
+                error_msg = error_msg + (
+                    'Following exploration ids are private %s' % (
```

Comment:
Put a `. ` after each error message, to separate them.

Reply:
Done.